mesonbuild/interpreter/interpreter.py

```diff
@@ -3135,29 +3135,31 @@ def do_validate_within_subproject(norm):
                 inputtype = 'directory'
             else:
                 inputtype = 'file'
-            if InterpreterRuleRelaxation.ALLOW_BUILD_DIR_FILE_REFERENCES in self.relaxations and builddir in norm.parents:
+            if InterpreterRuleRelaxation.ALLOW_BUILD_DIR_FILE_REFERENCES in self.relaxations and path_starts_with(norm, self.environment.build_dir):
                 return
-            if srcdir not in norm.parents:
+
+            if not path_starts_with(norm, self.environment.source_dir):
```

Comment:
Commit message for this is no longer accurate. :D

"Introduce a new function" -> use our recently added function?